umbra-js/classes/KeyPair.js

```diff
@@ -47,24 +91,39 @@ class KeyPair {
       // Multiply curve's generator point by private key to get public key
       const publicKey = ec.g.mul(this.privateKeyHexSlim);
 
-      // Save various forms of public key
-      this.publicKeyHexCoords = {
+      // Save off public key as hex, other forms computed as getters
+      const publicKeyHexCoords = {
         x: pad32ByteHex(publicKey.getX().toString('hex')),
         y: pad32ByteHex(publicKey.getY().toString('hex')),
       };
-      this.publicKeyHex = `0x04${this.publicKeyHexCoords.x}${this.publicKeyHexCoords.y}`;
+      this.publicKeyHex = `0x04${publicKeyHexCoords.x}${publicKeyHexCoords.y}`;
     } else if (key.length === 132) {
-      // Save various forms of public key
+      // PUBLIC KEY
+      // Save off public key as hex, other forms computed as getters
       this.publicKeyHex = key;
-      this.publicKeyHexCoords = {
-        x: pad32ByteHex(this.publicKeyHexSlim.slice(0, 64)),
-        y: pad32ByteHex(this.publicKeyHexSlim.slice(64)),
-      };
+    } else if (key.length === 66 && isTxHash) {
+      // TRANSACTION HASH
+      return (async () => {
```

Comment:
This line effectively means that sometimes the constructor returns a Promise, while sometimes it returns an instantiated instance. To me, a constructor should always do the latter. But is this a common pattern in JS or something?

If I were implementing this, I would have left the constructor as non-asynchronous, and used something like a static async generator function to return an instance created from a txHash. What do you think?

Reply:
I doubt this is a common pattern as I haven't come across it yet, so it probably isn't the most elegant approach and I'm open to other solutions.

I'm not totally sure I follow your suggestion—are you saying to use a two step approach where you'd do something like this?

```javascript
const publicKey = await KeyPair.recoverPublicKeyFromTransaction(txHash);
const keyPairInsance = new KeyPair(publicKey);
```

Reply:
Yup, basically. Where the two lines above would live in a static factory function.

Reply:
Updated in fd5508b